prov/rxm/src/rxm_cq.c

```diff
@@ -1391,12 +1392,16 @@ void rxm_ep_do_progress(struct util_ep *util_ep)
 			} else {
 				ret = 1;
 			}
+			if (OFI_UNLIKELY(!(--rxm_ep->cq_eq_fairness)))
+				goto check_eq;
 		} else if (ret < 0 && (ret != -FI_EAGAIN)) {
 			if (ret == -FI_EAVAIL)
 				rxm_cq_read_write_error(rxm_ep);
 			else
 				rxm_cq_write_error_all(rxm_ep, ret);
 		} else {
+check_eq:
+			rxm_ep->cq_eq_fairness = rxm_cq_eq_fairness;
```

Comment:
It doesn't look right progressing the EQ only when we read a successful completion, but not an error.  I wonder if it would be better to rewrite this as:
```
ret = fi_cq_read(...)
if (ret < 0) {
    ...
} else if (ret < 0 && (ret != -FI_EAGAIN)) {
    ...
}
if (ret == -FI_EAGAIN || --ep->cq_eq_fairness <= 0) {
    ep->cq_eq_fairness = rxm_cq_eq_fairness;
    ...
}
```
I'm not even sure we need the ret == - FI_EAGAIN check.  Maybe starting the EQ progress every X reads of the CQ is sufficient.

Reply:
Thanks, good suggestion. I made the change but did leave FI_EAGAIN in so that is more closely maps to the  check as it works now.